[vm] Do not define all arithmetic ops as `Pure`

kuhar opened the pull request:
In MLIR, the `Pure` attribute entails `AlwaysSpeculatable`. This is not a good default for arithmetic VM ops because they are implemented in terms of C functions that may invoke Undefined Behavior (UB) on some inputs, e.g., on signed integer overflow in `vm.add.i32`:
```c
static inline int32_t vm_add_i32(int32_t lhs, int32_t rhs) { return lhs + rhs; }
```

Update LICM-based speculation LIT test for arithmetic ops.

Next, we should update the C implementation of some ops affected by this change and make them speculatable.

Issue: https://github.com/iree-org/iree/issues/11179

benvanik commented:
There are some fairly large implications here - we'll need to take a deeper look at the IR coming out and performance of it.

kuhar commented:
> There are some fairly large implications here - we'll need to take a deeper look at the IR coming out and performance of it.

Yep, and some of these should be fixable if we also willing change the C implementation.

kuhar commented on compiler/src/iree/compiler/Dialect/VM/IR/VMOps.td:
```diff
@@ -464,7 +471,7 @@ def VM_GlobalRefOp :
 }
 
-def VM_GlobalAddressOp : VM_PureOp<"global.address", [
+def VM_GlobalAddressOp : VM_TrivialOp<"global.address", [
```
This one is an accidental change

kuhar commented on compiler/src/iree/compiler/Dialect/VM/IR/VMOps.td:
```diff
@@ -1943,7 +1950,7 @@ def VM_SwitchF64Op : VM_SwitchFloatOp<F64, "switch.f64", VM_OPC_SwitchF64,
 }
 
-def VM_SwitchRefOp : VM_PureOp<"switch.ref", [
+def VM_SwitchRefOp : VM_TrivialOp<"switch.ref", [
```
This one is an accidental change

kuhar commented:
Rebased. Added a TODO to make more ops pure.

I also ran a local benchmark to confirm this doesn't have a noticeable impact on performance. I don't think the numbers below show anything but noise. (The iree build configuration was release + assertions.)
 
```
➜ ~/iree/build-relass/benchmark_suites/TFLite  
$ cat ./jakub_no_pure_vmvx.vmfb | iree-benchmark-module --module_file=- --entry_function=main --function_input=1x384xi32 --function_input=1x384xi32 --function_input=1x384xi32 --device=local-task --task_topology_group_count=64 --list-devices 
Reading module contents from stdin...2022-11-28T11:44:37-05:00
Running iree-benchmark-moduleRun on (128 X 2250 MHz CPU s)
CPU Caches:  L1 Data 32 KiB (x64)
  L1 Instruction 32 KiB (x64)
  L2 Unified 512 KiB (x64)  L3 Unified 16384 KiB (x16)Load Average: 5.70, 22.56, 41.10
***WARNING*** Library was built as DEBUG. Timings may be affected.-----------------------------------------------------------------------------------------
Benchmark                               Time             CPU   Iterations UserCounters...-----------------------------------------------------------------------------------------
BM_main/process_time/real_time      57425 ms      1013614 ms            1 items_per_second=0.0174141/s
➜ ~/iree/build-relass/benchmark_suites/TFLite  
$ cat ./jakub_no_pure_vmvx.vmfb | iree-benchmark-module --module_file=- --entry_function=main --function_input=1x384xi32 --function_input=1x384xi32 --function_input=1x384xi32 --device=local-task --task_topology_group_count=64 --list-devices
Reading module contents from stdin...2022-11-28T11:45:44-05:00
Running iree-benchmark-moduleRun on (128 X 2250 MHz CPU s)
CPU Caches:
  L1 Data 32 KiB (x64)
  L1 Instruction 32 KiB (x64)
  L2 Unified 512 KiB (x64)  L3 Unified 16384 KiB (x16)
Load Average: 10.02, 20.70, 39.13
***WARNING*** Library was built as DEBUG. Timings may be affected.
-----------------------------------------------------------------------------------------
Benchmark                               Time             CPU   Iterations UserCounters...
-----------------------------------------------------------------------------------------
BM_main/process_time/real_time      56696 ms      1007579 ms            1 items_per_second=0.017638/s
➜ ~/iree/build-relass/benchmark_suites/TFLite  
$ cat ./jakub_pure_vmvx.vmfb | iree-benchmark-module --module_file=- --entry_function=main --function_input=1x384xi32 --function_input=1x384xi32 --function_input=1x384xi32 --device=local-task --task_topology_group_count=64 --list-devices 
Reading module contents from stdin...
2022-11-28T11:49:19-05:00
Running iree-benchmark-module
Run on (128 X 2250 MHz CPU s)
CPU Caches:
  L1 Data 32 KiB (x64)
  L1 Instruction 32 KiB (x64)
  L2 Unified 512 KiB (x64)
  L3 Unified 16384 KiB (x16)
Load Average: 1.71, 12.09, 31.98
***WARNING*** Library was built as DEBUG. Timings may be affected.
-----------------------------------------------------------------------------------------
Benchmark                               Time             CPU   Iterations UserCounters...
-----------------------------------------------------------------------------------------
BM_main/process_time/real_time      56977 ms       997041 ms            1 items_per_second=0.0175508/s
➜ ~/iree/build-relass/benchmark_suites/TFLite  
$ cat ./jakub_pure_vmvx.vmfb | iree-benchmark-module --module_file=- --entry_function=main --function_input=1x384xi32 --function_input=1x384xi32 --function_input=1x384xi32 --device=local-task --task_topology_group_count=64 --list-devices
Reading module contents from stdin...
2022-11-28T11:50:25-05:00
Running iree-benchmark-module
Run on (128 X 2250 MHz CPU s)
CPU Caches:
  L1 Data 32 KiB (x64)
  L1 Instruction 32 KiB (x64)
  L2 Unified 512 KiB (x64)
  L3 Unified 16384 KiB (x16)
Load Average: 5.61, 11.37, 30.40
***WARNING*** Library was built as DEBUG. Timings may be affected.
-----------------------------------------------------------------------------------------
Benchmark                               Time             CPU   Iterations UserCounters...
-----------------------------------------------------------------------------------------
BM_main/process_time/real_time      56305 ms       990032 ms            1 items_per_second=0.0177603/s
```

kuhar commented:
@benvanik I requested a new review since the approval was from 10 days ago.
